// bitgo/dist/src/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @prettier
 */
var _ = require("lodash");
// re-export from here for backwards compatibility reasons
var environments_1 = require("./v2/environments");
exports.Environments = environments_1.Environments;
var bitcoinNetwork;
var rmgNetwork;
/**
 * Set the global Bitcoin network. Used for v1 only.
 *
 * @deprecated
 */
function setNetwork(network) {
    if (network === 'bitcoin') {
        bitcoinNetwork = 'bitcoin';
    }
    else {
        // test network
        bitcoinNetwork = 'testnet';
    }
}
exports.setNetwork = setNetwork;
/**
 * Get the global Bitcoin network. Used for v1 only.
 *
 * @deprecated
 */
function getNetwork() {
    return bitcoinNetwork;
}
exports.getNetwork = getNetwork;
/**
 * Get the global RMG network. Used for v1 only.
 *
 * @deprecated
 */
function getRmgNetwork() {
    return rmgNetwork;
}
exports.getRmgNetwork = getRmgNetwork;
/**
 * Set the global RMG network. Used for v1 only.
 *
 * @deprecated
 */
function setRmgNetwork(network) {
    rmgNetwork = network;
}
exports.setRmgNetwork = setRmgNetwork;
/**
 * Helper function to validate the input parameters to an SDK method.
 * Only validates for strings - if parameter is different, check that manually
 *
 * @deprecated
 * @param params dictionary of parameter key-value pairs
 * @param expectedParams list of expected string parameters
 * @param optionalParams list of optional string parameters
 * @param optionalCallback if callback provided, must be a function
 * @returns true if validated, throws with reason otherwise
 */
function validateParams(params, expectedParams, optionalParams, optionalCallback) {
    if (optionalParams === void 0) { optionalParams = []; }
    if (!_.isObject(params)) {
        throw new Error('Must pass in parameters dictionary');
    }
    expectedParams = expectedParams || [];
    expectedParams.forEach(function (expectedParam) {
        if (!params[expectedParam]) {
            throw new Error('Missing parameter: ' + expectedParam);
        }
        if (!_.isString(params[expectedParam])) {
            throw new Error('Expecting parameter string: ' + expectedParam + ' but found ' + typeof params[expectedParam]);
        }
    });
    optionalParams.forEach(function (optionalParam) {
        if (params[optionalParam] && !_.isString(params[optionalParam])) {
            throw new Error('Expecting parameter string: ' + optionalParam + ' but found ' + typeof params[optionalParam]);
        }
    });
    if (optionalCallback && !_.isFunction(optionalCallback)) {
        throw new Error('illegal callback argument');
    }
    return true;
}
exports.validateParams = validateParams;
