// bitgo/dist/src/v2/trading/affirmation.js
"use strict";
/**
 * @prettier
 * @hidden
 */
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 */
var Bluebird = require("bluebird");
var co = Bluebird.coroutine;
var AffirmationStatus;
(function (AffirmationStatus) {
    AffirmationStatus["PENDING"] = "pending";
    AffirmationStatus["OVERDUE"] = "overdue";
    AffirmationStatus["REJECTED"] = "rejected";
    AffirmationStatus["AFFIRMED"] = "affirmed";
    AffirmationStatus["FAILED"] = "failed";
    AffirmationStatus["CANCELED"] = "canceled";
})(AffirmationStatus = exports.AffirmationStatus || (exports.AffirmationStatus = {}));
var Affirmation = /** @class */ (function () {
    function Affirmation(affirmationData, bitgo, enterpriseId) {
        this.bitgo = bitgo;
        this.enterpriseId = enterpriseId;
        this.updateAffirmationData(affirmationData);
    }
    /**
     * Affirms a pending affirmation, authorizing the settlement
     * @param payload payload authorizing the movement of funds from a trading account
     * @param signature signature of the payload with the user key of the trading account
     * @param callback
     */
    Affirmation.prototype.affirm = function (payload, signature, callback) {
        var body = {
            payload: JSON.stringify(payload),
            signature: signature,
        };
        return this.updateStatus(AffirmationStatus.AFFIRMED, body, callback);
    };
    /**
     * Rejects a pending affirmation, cancelling the settlement
     * @param callback
     */
    Affirmation.prototype.reject = function (callback) {
        return this.updateStatus(AffirmationStatus.REJECTED, null, callback);
    };
    /**
     * Cancels a pending affirmation, cancelling the settlement
     * @param callback
     */
    Affirmation.prototype.cancel = function (callback) {
        return this.updateStatus(AffirmationStatus.CANCELED, null, callback);
    };
    Affirmation.prototype.updateStatus = function (status, body, callback) {
        var self = this;
        return co(function updateStatus() {
            var bodyWithStatus, url, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        bodyWithStatus = __assign({ status: status }, body);
                        url = self.bitgo.microservicesUrl("/api/trade/v1/enterprise/" + self.enterpriseId + "/account/" + self.partyAccountId + "/affirmations/" + self.id);
                        return [4 /*yield*/, self.bitgo
                                .put(url)
                                .send(bodyWithStatus)
                                .result()];
                    case 1:
                        response = _a.sent();
                        self.updateAffirmationData(response);
                        return [2 /*return*/];
                }
            });
        })
            .call(this)
            .asCallback(callback);
    };
    Affirmation.prototype.updateAffirmationData = function (affirmationData) {
        this.id = affirmationData.id;
        this.partyAccountId = affirmationData.partyAccountId;
        this.status = affirmationData.status;
        this.settlement = affirmationData.settlement;
        this.lock = affirmationData.lock;
        this.payload = affirmationData.payload;
        this.createdAt = new Date(affirmationData.createdAt);
        this.expireAt = new Date(affirmationData.expireAt);
    };
    return Affirmation;
}());
exports.Affirmation = Affirmation;
