// @bitgo/account-lib/dist/src/coin/baseCoin/errors.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This base class ensures that our stack trace is captured properly but also that we have classes of errors
 * that can be found in a switch.
 */
var ExtendableError = /** @class */ (function (_super) {
    __extends(ExtendableError, _super);
    function ExtendableError(message) {
        var _this = _super.call(this, message) || this;
        _this.name = _this.constructor.name;
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(_this, _this.constructor);
        }
        else {
            _this.stack = (new Error(message)).stack;
        }
        return _this;
    }
    return ExtendableError;
}(Error));
exports.ExtendableError = ExtendableError;
var ParseTransactionError = /** @class */ (function (_super) {
    __extends(ParseTransactionError, _super);
    function ParseTransactionError(message) {
        return _super.call(this, message) || this;
    }
    return ParseTransactionError;
}(ExtendableError));
exports.ParseTransactionError = ParseTransactionError;
var SigningError = /** @class */ (function (_super) {
    __extends(SigningError, _super);
    function SigningError(message) {
        return _super.call(this, message) || this;
    }
    return SigningError;
}(ExtendableError));
exports.SigningError = SigningError;
var AddressValidationError = /** @class */ (function (_super) {
    __extends(AddressValidationError, _super);
    function AddressValidationError(message) {
        return _super.call(this, message) || this;
    }
    return AddressValidationError;
}(ExtendableError));
exports.AddressValidationError = AddressValidationError;
var BuildTransactionError = /** @class */ (function (_super) {
    __extends(BuildTransactionError, _super);
    function BuildTransactionError(message) {
        return _super.call(this, message) || this;
    }
    return BuildTransactionError;
}(ExtendableError));
exports.BuildTransactionError = BuildTransactionError;
var UtilsError = /** @class */ (function (_super) {
    __extends(UtilsError, _super);
    function UtilsError(message) {
        return _super.call(this, message) || this;
    }
    return UtilsError;
}(ExtendableError));
exports.UtilsError = UtilsError;
var ExtendTransactionError = /** @class */ (function (_super) {
    __extends(ExtendTransactionError, _super);
    function ExtendTransactionError(message) {
        return _super.call(this, message) || this;
    }
    return ExtendTransactionError;
}(ExtendableError));
exports.ExtendTransactionError = ExtendTransactionError;
